backend/src/main/kotlin/org/icpclive/cds/ejudge/EjudgeDataSource.kt

```diff
@@ -0,0 +1,281 @@
+            }
+        }
+    }
+
+    private fun parseRuns(
+        contestInfo: EjudgeContestInfo,
+        element: Element,
+        onRunChanges: (RunInfo) -> Unit
+    ) {
+        if (contestInfo.status == ContestStatus.BEFORE) {
+            return
+        }
+        element.children().forEach { run ->
+            parseRunInfo(contestInfo, run, onRunChanges)
+        }
+    }
+
+    private fun parseRunInfo(
+        contestInfo: EjudgeContestInfo,
+        element: Element,
+        onRunChanges: (RunInfo) -> Unit
+    ) {
+        val time = (element.attr("time").toLong() * 1000).milliseconds
+        if (time > contestInfo.contestTime) {
+            return
+        }
+
+        val teamSystemId = element.attr("user_id").toInt()
+        val teamId = contestInfo.teams[teamSystemId.toString()]!!.teamInfo.id
+        val runId = element.attr("run_id").toInt()
+
+        // Ejudge has 1-indexed problem numeration
+        val problemId = element.attr("prob_id").toInt() - 1
+        val isFrozen = time >= contestInfo.freezeTime
+        val oldRun = contestInfo.teams[teamSystemId.toString()]!!.runs[problemId].getOrDefault(runId, null)
+        val result = when {
+            isFrozen -> ""
+            else -> statusMap.getOrDefault(element.attr("status"), "WA")
+        }
+        val percentage = when {
+            isFrozen -> 0.0
+            "" == result -> 0.0
+            else -> 1.0
+        }
+
+        val run = RunInfo(
+            id = oldRun?.id ?: element.attr("run_id").toInt(),
+            isAccepted = "AC" == result,
+            isJudged = "" != result,
+            isAddingPenalty = "AC" != result && "CE" != result,
+            result = result,
+            problemId = problemId,
+            teamId = teamId,
+            percentage = percentage,
+            time = time.inWholeMilliseconds.toDuration(DurationUnit.MILLISECONDS),
+            isFirstSolvedRun = false
+        )
+
+        if (run != oldRun) {
+            onRunChanges(run)
+        }
+        contestInfo.teams[teamSystemId.toString()]!!.runs[problemId][runId] = run
+    }
+
+    private var contestData: EjudgeContestInfo
+    private var contestInfoFlow: MutableStateFlow<ContestInfo>
+    private val properties: Properties = Config.loadProperties("events")
+    private val xmlLoader = object : RegularLoaderService<Document>(null) {
+        override val url = properties.getProperty("url")
+        override fun processLoaded(data: String) = Jsoup.parse(data, "", Parser.xmlParser())
+    }
+
+    init {
+        val problemsInfo = parseProblemsInfo()
+        val teamsInfo = parseTeamsInfo(problemsInfo.size)
+
+        contestData = EjudgeContestInfo(
+            problemsInfo,
+            teamsInfo.associateBy { it.teamInfo.contestSystemId },
+            Instant.fromEpochMilliseconds(0),
+            ContestStatus.UNKNOWN,
+            properties.getProperty("contest.length")?.toInt()?.milliseconds ?: 5.hours,
```

Comment:
maybe this data exists in external.xml? It would be much better to retrieve it from it. In PCMS it's a hack, which is required due to lack of information